waspc/src/Wasp/Generator/NpmInstall.hs

```diff
@@ -100,34 +109,81 @@ loadInstalledFullStackNpmDependencies dstDir = do
+        "You've been waiting so patiently, just wait a little longer (for the installation to finish)..."
+      ]
+
+installNpmDependenciesAndReport :: Job -> Chan JobMessage -> JobType -> IO ExitCode
+installNpmDependenciesAndReport installF chan jobType = do
+  writeChan chan $ J.JobMessage {J._data = J.JobOutput "Starting npm install\n" J.Stdout, J._jobType = jobType}
+  result <- installF chan `race` reportInstallationProgress chan jobType
+  case result of
+    Left exitCode -> return exitCode
+    Right _ -> error "This should be impossible"
+
 -- Run the individual `npm install` commands for both server and webapp projects
 -- It runs these concurrently, collects the output produced by these commands
 -- to pass them along to IO with a prefix
-installNpmDependencies :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
-installNpmDependencies projectDir = do
-  chan <- newChan
-  let runSetupJobs =
-        ServerSetup.installNpmDependencies projectDir chan
-          `concurrently` WebAppSetup.installNpmDependencies projectDir chan
-  (_, result) <- concurrently (handleJobMessages chan) runSetupJobs
-  case result of
-    (ExitSuccess, ExitSuccess) -> return $ Right ()
-    exitCodes -> return $ Left $ setupFailedMessage exitCodes
+installNpmDependencies :: Path' Abs (Dir WaspProjectDir) -> Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
+installNpmDependencies projectDir dstDir = do
+  messagesChan <- newChan
+  (_, exitCode) <-
+    concurrently
+      (handleProjectInstallMessage messagesChan)
+      (installNpmDependenciesAndReport (SdkGenerator.installNpmDependencies projectDir) messagesChan J.Wasp)
+  case exitCode of
+    ExitFailure code -> return $ Left $ "Project setup failed with exit code " ++ show code ++ "."
+    _ -> do
+      let handleMessagesJob = handleJobMessages messagesChan
+      let runSetupJobs =
+            concurrently
+              (installNpmDependenciesAndReport (ServerSetup.installNpmDependencies dstDir) messagesChan J.Server)
+              (installNpmDependenciesAndReport (WebAppSetup.installNpmDependencies dstDir) messagesChan J.WebApp)
+      (_, results) <- concurrently handleMessagesJob runSetupJobs
+      case results of
+        (ExitSuccess, ExitSuccess) -> return $ Right ()
+        exitCodes -> return $ Left $ setupFailedMessage exitCodes
   where
-    handleJobMessages = runPrefixedWriter . go (False, False)
+    handleProjectInstallMessage :: Chan J.JobMessage -> IO ()
+    handleProjectInstallMessage = runPrefixedWriter . processMessages
+      where
+        processMessages :: Chan J.JobMessage -> PrefixedWriter ()
+        processMessages chan = do
+          jobMsg <- liftIO $ readChan chan
+          case J._data jobMsg of
+            J.JobOutput {} -> printJobMessagePrefixed jobMsg >> processMessages chan
+            J.JobExit {} -> return ()
+    handleJobMessages = runPrefixedWriter . processMessages (False, False)
```

Comment:
We have `handleProjectInstallMessage` which is specific, but then `handleJobMessages` name is generic. It should also be specific instead, something like `handleSetupJobsMessages` or whatever you think describes well the two jobs that install server and web app.

Reply:
Yes, I'll do that.

I kinda forgot to do my final self-review pass on this PR, as you probably noticed.

Reply:
No worries, haven't really noticed!